### DISCUSSION TITLE
Abstraction Between Blazor Server and CSLA

### DISCUSSION THREAD

**Discussion by jmpotvin** [MarimerLLC/csla#3555] [Questions]
I'm integrating CSLA with a Blazor Server frontend and considering adding an abstraction layer between them for better separation and scalability. I would appreciate insights on:

1. Is this a common/recommended practice in CSLA implementations?
2. Why is there no specific project or template for such an abstraction? Are there known drawbacks?

Any guidance or references to similar setups would be greatly helpful.

**Comment by rockfordlhotka** [comment]
An abstraction layer between what and what? What is "them" in this context?

**Reply by jmpotvin** [reply]
The CSLA business logic and the Blazor Server frontend.

**Reply by rockfordlhotka** [reply]
I'm not entirely sure how you'd do such a think without losing most of the value of CSLA.

I think what you are asking is to build a client app that _doesn't_ bind to the domain types, but rather keeps all the domain types (for example) behind a service API.

That's certainly the model people use when building an Angular or React (or whatever) app that calls services, where the services are implemented using CSLA.

The problem with that, is that all the business logic in C# (CSLA) is running on the app server, so the user experience is horrible. To address that problem, much (most?) of the business logic must be reimplemented in TypeScript/JavaScript so it can also run on the client to give the user a good experience. Then the dev team must manually ensure that the frontend and backend code remain in sync via code reviews or some other inefficient process.

You surely _could_ do that with Blazor, by reimplementing all the business logic in the Blazor UI and manually keeping all that logic in sync between the UI code and app server code.

That flies in the face of the intent of CSLA, which is to allow you to write and maintain the business logic in one set of business domain classes, and to deploy the DLL containing that code to client and server.

It also prevents leveraging all the powerful capabilities around CSLA that support data binding, and the APIs that expose business and authorization rules for use in building a highly interactive user experience.

**Comment by jmpotvin** [comment]
I'd like to suggest the idea of developing a Csla.Blazor.Abstractions project. This project would incorporate existing interfaces such as IPropertyInfo, and potentially introduce new ones to support various types of business objects. The goal isn't to duplicate logic or change CSLA's core functionality but to create a stable interface between CSLA business logic and Blazor frontends. Key points to consider:

1. Change Resilience: These interfaces could shield apps from changes in CSLA versions, enhancing adaptability.
2. Testing Efficiency: This approach would make it easier to mock components for unit tests, leading to more reliable and isolated testing scenarios.
3. Clearer Architecture: The implementation of these interfaces would clarify the separation between CSLA logic and Blazor UI, aiding in maintenance and scalability.

The intention is to establish a well-defined interface layer for safer and more maintainable CSLA interactions.

Looking forward to your thoughts on this.

**Reply by Chicagoan2016** [reply]
If I am following correctly, wouldn't it result in re-implementing the same interfaces (in Csla.Blazor.Abstractions project) already implemented by Csla?

**Comment by jmpotvin** [comment]
Regarding the Csla.Blazor.Abstractions idea and looking at the CSLA BlazorServerExample, the focus should probably on creating interfaces for Csla.Blazor.ViewModel and Csla.Blazor.ViewModel.Model. This approach would allow projects to depend on Csla.Blazor.Abstractions instead of directly on Csla.Blazor, offering these advantages:

1. Less Version Lock-In: Reduces binding to specific Csla.Blazor versions, aiding in future-proofing applications.
2. Enhanced Unit Testing: Abstracting these components would allow for more effective mocking in unit tests, improving testability.
3. Easier Maintenance: This approach could simplify updates and maintenance as Csla.Blazor evolves.

Any views on this?

**Comment by jmpotvin** [comment]
Answered by @rockfordlhotka here:
https://github.com/MarimerLLC/csla/discussions/3610#discussioncomment-7846725

**Comment by jmpotvin** [comment]
Answered by @rockfordlhotka here:
https://github.com/MarimerLLC/csla/discussions/3610#discussioncomment-7846725
